File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> res(nums.size(), -1);
        deque<int> myDeque;
        myDeque.push_back(0);

        for (int i = 1; i < nums.size() * 2; i++) {
            int idx = i % nums.size();
            if(nums[idx] <= nums[myDeque.back()]) {
                myDeque.push_back(idx);
            } else {
                while(!myDeque.empty() && nums[idx] > nums[myDeque.back()]) {
                    res[myDeque.back()] = nums[idx];
                    myDeque.pop_back();
                }
                myDeque.push_back(idx);
            }
        }
        return res;
    }
};

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        res = [-1] * len(nums)

        nums = nums + nums

        stack = [0]

        for i in range(1, len(nums)):
            while stack and nums[i] > nums[stack[-1]]:
                idx = stack.pop() % len(res)
                res[idx] = nums[i]

            stack.append(i)

        return res
